Various Jaunt fixes

Jacquerel opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request

This reworks the flow of how Jaunt spells work somewhat to fix a number of bugs I found while fucking around with them downstream.

Fixes https://github.com/tgstation/tgstation/issues/56254
Fixes https://github.com/tgstation/tgstation/issues/49715
Fixes https://github.com/tgstation/tgstation/issues/46620
As well as some unreported bugs I found which were chiefly related to `TRAIT_MAGICALLY_PHASED` persisting after being forced to exit the jaunt when you didn't do it voluntarily.

The mob holder object which does the funny jaunt movement now sends `COMSIG_MOB_EJECTED_FROM_JAUNT` when the mob exits its contents.
The jaunt spell registers to this and calls new proc `on_jaunt_exited` when it receives it. This is where any removal of traits or effects should go, because it will be called any time the mob is ejected rather than only if they did it on purpose.

This also means that the spell will clean itself up more effectively if it is removed from you while it is active, for instance in the above linked issue where someone was ahealed and it ejected the demon heart which gave them the spell.

Additionally, the mob procs to move up or down will now call `relayMove` instead if you are inside an object or mob. I expected this to be more complicated than that, but that seems to have just fixed the issue immediately and now if you are capable of z-travelling you can still do it while jaunted without moving yourself out of the jaunt object and ending it prematurely.

Additionally Mirror Walk's button was fucked up and would display itself as active or inactive arbitrarily, depending on whether you were near a mirror the last time the cooldown ended. Now it also checks this whenever you move, so it will correctly live update. I suspect nobody reported this because the button is already red so tinting it red actually isn't very obvious.
While I was at it, I added the same functionality to Blood Crawl and Shadow Walk for better visual feedback.

## Why It's Good For The Game

It fixes some bugs, most of which were only available in niche circumstances.
Z travelling while jaunting is pretty useful on Icebox and Tram for a Maid in the Mirror and it was weird and sometimes problematic that it would instead just reveal you, which could give away information you didn't intend.
The mirror walk button cooldown effect was fucked up and now it isn't.
Feedback on whether blood crawl or shadow walk will actually do anything before you press it is more user friendly.

## Changelog

<!-- If your PR modifies aspects of the game that can be concretely observed by players or admins you should add a changelog. If your change does NOT meet this description, remove this section. Be sure to properly mark your PRs to prevent unnecessary GBP loss. You can read up on GBP and it's effects on PRs in the tgstation guides for contributors. Please note that maintainers freely reserve the right to remove and add tags should they deem it appropriate. You can attempt to finagle the system all you want, but it's best to shoot for clear communication right off the bat. -->

:cl:
fix: Changing z-level while jaunting will no longer reveal you.
fix: If a shuttle lands on you while jaunting you will die rather than be stuck in a weird purgatory state.
fix: Mirror Walk will not appear tinted as if it can be cast when it can't, and vice versa.
qol: Blood Crawl and Shadow Walk will now display whether you can use them or not in the same manner as Mirror Walk.
refactor: Jaunt code which adds and removes temporary traits or effects should hopefully be easier to work with in the future.
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

Jacquerel committed:
Jaunt code path reworked so that traits and other effects can be removed consistently regardless of how effect is ended.
Jaunts will more consistently clean themselves up (and unjaunt you) when you lose the spell.
If a shuttle lands on you while you are jaunted it will now kill you instead of crashing and fucking with the shuttle landing process.
Z travelling while inside an object or mob will now relay that direction instead, allowing you to jaunt up and down as well as cardinally.
Mirror walk button updates at correct times.
Blood and Shadow walk buttons have same functionality as Mirror Walk.

MrMelbert reviewed (changes requested):
I was looking into doing multi-z jaunt support a while back, cool stuff

You should probably be careful about giving Ethereal Jaunt and Shadow Walk full multi-z support, though - I personally think it should be limited in some way, otherwise, a wizard or nightmare could be nigh un-killable by simply moving downwards or upwards when they trigger their jaunts

Some ideas could include...
- Disallowing jaunts into non-station areas
- Require a do_after for vertical movement
- Telegraph that a vertical movement occurred somehow
- Force them out of the jaunt shortly afterwards, maybe with a short stun

MrMelbert reviewed (commented):
Also, I don't know if there's any side-effects from doing `setDir(UP)` or `setDir(DOWN)`, but `/phased_mob` will do that in relaymove

MrMelbert reviewed (commented):
It is also mindful to update `/obj/effect/dummy/phased_mob/proc/phased_check(mob/living/user, direction)` to accommodate - using `get_step_multiz` instead of just `get_step`

Jacquerel committed:
Minor refactors to blood crawl efficiency.
Don't set your dir to UP or DOWN.
Jaunting up and down now works as described in my really long comment.

Jacquerel commented:
Following that push now works as described above, IE:
Pressing "move up or down" while above a floor or below a ceiling doesn't do anything.
Pressing "move up or down" with an open space above or below you will move into that space (allowing you to fly even if you cannot usually).
Doing this does not forcemove you out of the jaunt object so you continue jaunting from this point.

Mothblocks commented on code/game/objects/effects/phased_mob.dm:
```diff
@@ -32,7 +32,7 @@
 /// Removes [jaunter] from our phased mob
 /obj/effect/dummy/phased_mob/proc/eject_jaunter()
 	if(!jaunter)
-		CRASH("Phased mob ([type]) attempted to eject null jaunter.")
```
Give me a chance today to look at how many times this runtime has occurred, and where, location wise, it has happened. I want to make sure we're not going to be hiding bugs.

Mothblocks replied:
It's happened 0 times in the entire year of 2022, you're safe

Mothblocks commented on code/modules/antagonists/heretic/magic/mirror_walk.dm:
```diff
@@ -88,21 +100,25 @@
+		return
+	unjaunter.visible_message(
+		span_boldwarning("[unjaunter] phases into reality before your very eyes!"),
+		span_notice("You jump out of the reflection coming off of [nearby_reflection], exiting the mirror's realm."),)
```
```suggestion
		span_notice("You jump out of the reflection coming off of [nearby_reflection], exiting the mirror's realm."),
	)
```

Mothblocks commented on code/modules/mob/mob_movement.dm:
```diff
@@ -529,6 +529,10 @@
 		return
 
+	if(ismovable(loc)) //Inside an object, tell it we moved
+		var/atom/O = loc
+		return O.relaymove(src, UP)
```
```suggestion
		var/atom/loc_atom = loc
		return loc_atom.relaymove(src, UP)
```

Mothblocks commented on code/modules/mob/mob_movement.dm:
```diff
@@ -544,6 +548,10 @@
 	set category = "IC"
 
+	if(ismovable(loc)) //Inside an object, tell it we moved
+		var/atom/O = loc
+		return O.relaymove(src, DOWN)
```
```suggestion
		var/atom/loc_atom = loc
		return loc_atom.relaymove(src, DOWN)
```

Jacquerel committed:
Rename copy/pasted single letter vars.
Minor format change.
